flow/Deque.h

```diff
@@ -41,21 +41,28 @@ class Deque {
 	Deque() : arr(0), begin(0), end(0), mask(-1) {}
 
 	// TODO: iterator construction, other constructors
-	Deque(Deque const& r) : arr(0), begin(0), end(r.size()), mask(r.mask) {
+	Deque(Deque const& r) : arr(nullptr), begin(0), end(r.size()), mask(r.mask) {
 		if (r.capacity() > 0) {
 			arr = (T*)aligned_alloc(std::max(__alignof(T), sizeof(void*)), capacity() * sizeof(T));
 			ASSERT(arr != nullptr);
 		}
 		ASSERT(capacity() >= end || end == 0);
-		for (uint32_t i=0; i<end; i++)
-			new (&arr[i]) T(r[i]);
-		// FIXME: Specialization for POD types using memcpy?
+		if (r.end < r.capacity()) {
+			std::copy(r.arr + r.begin, r.arr + r.begin + r.size(), arr);
+		} else {
+			// Mask is used for wrapping, so r.end is always >= r.begin,
+			// but if r.end >= r.capacity(), the deque wraps around so the
+			// copy must be performed in two parts
+			auto partOneSize = r.capacity() - r.begin;
+			std::copy(r.arr + r.begin, r.arr + r.begin + partOneSize, arr);
```

Comment:
```suggestion
			auto partTwo = std::copy(r.arr + r.begin, r.arr + r.capacity(), arr);
```

Reply:
These changes to the std::copy calls are a matter of taste; feel free to ignore.  They eliminate the partOneSize arithmetic (and can be applied to the similar logic below, if you want).